lua/libcairo_text_helper.h

```diff
@@ -0,0 +1,281 @@
+  hb_direction_t text_direction = hb_direction_from_string(direction, -1);
+  if (text_direction == HB_DIRECTION_INVALID) {
+    text_direction = HB_DIRECTION_LTR;
+  }
+  hb_script_t text_script = hb_script_from_string(script, -1);
+  if (text_script == HB_SCRIPT_UNKNOWN) {
+    text_script = HB_SCRIPT_COMMON;
+  }
+    
+  hb_language_t text_language = hb_language_from_string (language, -1);
+  
+  /* Draw text */
+  /* Create a buffer for harfbuzz to use */
+  hb_buffer_t *buf = hb_buffer_create();
+
+  //alternatively you can use hb_buffer_set_unicode_funcs(buf, hb_glib_get_unicode_funcs());
+  hb_buffer_set_unicode_funcs(buf, hb_glib_get_unicode_funcs());
+  hb_buffer_set_direction(buf, text_direction); /* or LTR */
+  hb_buffer_set_script(buf, text_script); /* see hb-unicode.h */
+  hb_buffer_set_language(buf, text_language);
+
+  /* Layout the text */
+  hb_buffer_add_utf8(buf, text, strlen(text), 0, strlen(text));
+  hb_shape(font->hb_ft_font, buf, NULL, 0);
+
+  /* Hand the layout to cairo to render */
+  unsigned int         glyph_count;
+  hb_glyph_info_t     *glyph_info   = hb_buffer_get_glyph_infos(buf, &glyph_count);
+  hb_glyph_position_t *glyph_pos    = hb_buffer_get_glyph_positions(buf, &glyph_count);
+  cairo_glyph_t       *cairo_glyphs = malloc(sizeof(cairo_glyph_t) * glyph_count);
+
+  double x1=0, x2=0, y1=0, y2=0;
+  cairo_clip_extents(cr, &x1, &y1, &x2, &y2);
+  int width = floor(x2-x1);
+  int height = floor(y2-y1);
+
+  unsigned int string_width_in_pixels = 0;
+  for (int i=0; i < glyph_count; ++i) {
+      string_width_in_pixels += glyph_pos[i].x_advance/64;
+  }
+  int draw_x = x;
+
+  if (HB_DIRECTION_IS_VERTICAL(text_direction)) { 
+    /* FIXME */
+    draw_x = width/2 - string_width_in_pixels/2 + x;
+  }
+  else {
+    if (alignment == CAIRO_TEXT_ALIGN_LEFT) {
+      if (text_direction == HB_DIRECTION_RTL) {
+        draw_x = width - x - string_width_in_pixels;
+      }
+      // LTR handled as default.
+    }
+    else if (alignment == CAIRO_TEXT_ALIGN_RIGHT) {
+      if (text_direction == HB_DIRECTION_RTL) {
+        draw_x = x;
+      }
+      else {
+        draw_x = width - x - string_width_in_pixels;
+      }
+    }
+    else if (alignment == CAIRO_TEXT_ALIGN_CENTER) {
+      draw_x = width/2 - string_width_in_pixels/2 + x;
+    }
+  }
+
+  x = draw_x;
+  for (int i=0; i < glyph_count; ++i) {
+      cairo_glyphs[i].index = glyph_info[i].codepoint;
+      cairo_glyphs[i].x = x + (glyph_pos[i].x_offset/64.0);
+      cairo_glyphs[i].y = y - (glyph_pos[i].y_offset/64.0);
+      x += glyph_pos[i].x_advance/64.0;
+      y -= glyph_pos[i].y_advance/64.0;
+  }
+  
+  cairo_set_source_rgba (cr, 1.0, 1.0, 1.0, 1.0);
+  cairo_set_font_face(cr, font->cairo_ft_face);
+  cairo_set_font_size(cr, font->font_size);
+  cairo_show_glyphs(cr, cairo_glyphs, glyph_count);
+
+  free(cairo_glyphs);
+  hb_buffer_destroy(buf);
+}
+
+void cairo_text_hp_simple_show(cairo_t *cr, int x, int y, const char *text, FontData *font)
```

Comment:
Note that there's no need for these `cairo_text_hp_simple_show_*` convenience functions. toluapp supports default values and harfbuzz can infer defaults too.

You can make all of the properties optional by:
- adding a call to `hb_buffer_guess_segment_properties (hb_buffer);` in `cairo_text_hp_intl_show` after _specified properties_ are set.
- setting argument order to be: `language` (because it's inferred from system (least smart)), `script` (inferred from buffer contents), `direction` (inferred from script)
  - as documented [here](https://harfbuzz.github.io/harfbuzz-hb-buffer.html#hb-buffer-guess-segment-properties)
  - that order would minimize the need for supplying well inferred values to set the one without a good default
- specifying argument defaults to be `nullptr`/`NULL` in the header
  - and not calling `hb_*_from_string` if the value is null (i.e. empty string).

Reply:
In the same line of reasoning I'd position cairo `alignment` argument after the `font` argument and set it to default to `CAIRO_TEXT_ALIGN_LEFT`.

Reply:
Also, the binding function should be able to load and cache a font. The cache can leak memory (no need to call free) as it would be weird to manage such things from a lua script and fonts will likely be needed for each draw call.

So the signature of `cairo_text_hp_intl_show` in .pkg would finally look like: `void cairo_text_hp_intl_show(cairo_t *cr, int x, int y, const char *text, const char *font = NULL, int font_size = 12, cairo_text_alignment_t alignment = CAIRO_TEXT_ALIGN_LEFT, const char *language = NULL, const char *script = NULL, const char *direction = NULL)`.

Reply:
Sorry I didn't see this until now, I'll certainly swap to the idea of default arguments that seems much better. Leaking fonts seems reasonable to me as well so i'm happy to just add a cache.

Reply:
I am implementing most of this, with the exception of `language` which i'm going to default to english if not explicitly set. The reason being if I create a config / script for conky and want to distribute it, I don't want it to start breaking in random ways because other people are using a different language. Also much of conky's outputs such as process names CPU percentages etc work best in English. In this case if you are doing something that requires an explicit language you'll need to set it as well.